src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -0,0 +1,1297 @@
+         * @param costCoefficients controls the relative costs of various database operations
+         */
+        public Factory(long totalRows, CostCoefficients costCoefficients)
+        {
+            this.totalRows = totalRows;
+            this.costCoefficients = costCoefficients;
+            this.nothing = new Nothing(-1, this);
+            this.everything = new Everything(-1, this);
+        }
+
+        /**
+         * Constructs a plan node representing a direct scan of an index.
+         * @param matchingKeysCount the number of row keys expected to be returned by the index scan,
+         *                        i.e. keys of rows that match the search predicate
+         * @param keyIterator the factory for creating the iterator over matching keys
+         */
+        public KeysIteration indexScan(@Nonnull String description,
+                                       long matchingKeysCount,
+                                       @Nonnull Supplier<RangeIterator> keyIterator)
+        {
+            Preconditions.checkNotNull(keyIterator, "keyIterator supplier must not be null");
+            Preconditions.checkArgument(matchingKeysCount >= 0, "matchingKeyCount must not be negative");
+            Preconditions.checkArgument(matchingKeysCount <= totalRows, "matchingKeyCount must not exceed totalKeyCount");
+            return new IndexScan(this, nextId++, description, matchingKeysCount, keyIterator);
+        }
+
+        /**
+         * Constructs a plan node representing a union of two key sets.
+         * @param subplans a list of subplans for unioned key sets
+         */
+        public KeysIteration union(List<KeysIteration> subplans)
+        {
+            return union(subplans, nextId++);
+        }
+
+        private KeysIteration union(List<KeysIteration> subplans, int id)
+        {
+            if (subplans.contains(everything))
+                return everything;
+            if (subplans.contains(nothing))
+                subplans.removeIf(s -> s == nothing);
+            if (subplans.size() == 1)
+                return subplans.get(0);
+            if (subplans.isEmpty())
+                return nothing;
+
+            return new Union(this, id, subplans);
+        }
+
+        /**
+         * Constructs a plan node representing an intersection of two key sets.
+         * @param subplans a list of subplans for intersected key sets
+         */
+        public KeysIteration intersect(List<KeysIteration> subplans)
+        {
+            return intersect(subplans, nextId++);
+        }
+
+        private KeysIteration intersect(List<KeysIteration> subplans, int id)
+        {
+            if (subplans.contains(nothing))
+                return nothing;
+            if (subplans.contains(everything))
+                subplans.removeIf(c -> c == everything);
+            if (subplans.size() == 1)
+                return subplans.get(0);
+            if (subplans.isEmpty())
+                return everything;
```

Comment:
This is interesting, the intersection of "nothing to keep" is "everything" ?

Reply:
Yes. The zero element of a multiplication is 1. And generally it is a zero element of the WHERE clause restrictions algebra.

E.g. `SELECT * FROM tab WHERE x < 5` selects a subset of rows, but `SELECT * FROM tab` (no WHERE) select *all* rows. 

That simplifies a lot of stuff, e.g. we can freely remove restrictions anywhere from the tree without making a correctness error, because returning false positives from the index is always ok (there is post-filtering).